### PR TITLE
Feat branch complete on attach

### LINKED ISSUE #1475: [Followup] full migration to on_attach API mappings
#1579

- [x] document default mappings
- [x] apply `remove_keymaps`
- [x] legacy migration #1464
- [x] help shows on_attach mappings
- [ ] update help
- [ ] write migration guide
- [ ] ensure `action=""` applied to migration

### PR DESCRIPTION
**Pull request by alex-courtis**
Feature branch containing all functionality to migrate from mappings to on_attach keymaps.

Resolves #1475

New
- [x] `chore/action_cb_help_poc` help displays vim mappings with their description
- [x] :help
- [ ] ~~migration notification (yes, some users will get upset)~~ no need at this time
- [x] wiki
- [x] apply default mappings when no `on_attach`

Legacy
- [x] `chore/migrate-legacy-mappings` legacy migration of mappings to an on_attach function
- [x] `chore/migrate-legacy-mappings-via-keymap` remove mappings and dispatcher
- [x] `chore/migrate-legacy-mappings-tool` :NvimTreeGenerateOnAttach
- [x] `action = ""` translated to `remove_keymaps`
- [x] ~~warn when `remove_keymap` or `action = ""` key not found as this has been consistently problematic RE case etc.~~ silently handle removal of inexistent mappings
- [x] comment on removed keymap pattern
- [x] opts helper when mapping default and generated

### CODE REVIEW

**Comment by kyazdani42** [comment]
@alex-courtis sorry i haven't been much here lately, i've been kind of tired to do anything related to IT to be honest.
Thank you for managing the plugin while i'm not here. I'm not sure when i'll be back to work on any of my github projects, if you feel like you don't want to maintain it alone in the meantime, i can try to find another maintainer, just tell me !

**Comment by alex-courtis** [comment]
> @alex-courtis sorry i haven't been much here lately, i've been kind of tired to do anything related to IT to be honest. Thank you for managing the plugin while i'm not here.

Glad to hear you are back and not intending to stress yourself.

> I'm not sure when i'll be back to work on any of my github projects, if you feel like you don't want to maintain it alone in the meantime, i can try to find another maintainer, just tell me !

Another maintainer would be great. There is a regular workload of bugs and PRs. I am able to dedicate time now, but that may change.

Perhaps we look at some PRs and see if anyone stands out that we could approach.

@gegoune : your ideas and different approaches to functionality have been most insightful. Perhaps you could be a maintainer, with a mandate of product management over development.

**Comment by gegoune** [comment]
I would be very happy to help to full extend of my abilities. Thank you for kind words @alex-courtis! ❤️

**Comment by alex-courtis** [comment]
Dogfooding now following successful `:NvimTreeGenerateOnAttach` for my config.

**Comment by alex-courtis** [comment]
Testing 00be1db

* actual mappings
* help
* `:NvimTreeGenerateOnAttach`
* `api.config.mappings.default`, `.active` against master

|  | actual | help | generated | api |
| - | - | - | - | - |
| defaults  | OK  | OK | defaults | OK |

```lua
local function print_node_path(node)
  print(node.absolute_path)
end

require("nvim-tree").setup({
  -- remove_keymaps = true,
  -- remove_keymaps = {
  --   -- inexistent
  --   "X",
  --   -- exists
  --   "Y",
  -- },
  view = {
    mappings = {
      -- custom_only = true,
      -- list = {
      --   -- new
      --   { key = "A", action = "cd" },
      --   { key = "?", action = "toggle_help" },

      --   -- override
      --   { key = "B", action = "cd" },

      --   -- remove existing
      --   { key = { "<2-RightMouse>" }, action = "" },
      --   { key = "C", action = "" },
      --   { key = { "D", "E" }, action = "" },

      --   -- remove inexistent
      --   { key = "G", action = "" },
      --   { key = { "L", "M" }, action = "" },

      --   -- custom overwrite
      --   { key = "F", action = "print_the_node_path", action_cb = print_node_path },

      --   -- custom new
      --   { key = { "N", "Q" }, action = "print_the_node_path", action_cb = print_node_path },
      -- },
    },
  },
  log = {
    enable = true,
    truncate = true,
    types = {
      dev = true,
    },
  },
})

local api = require("nvim-tree.api")
local log = require("nvim-tree.log")

local default = api.config.mappings.default()
for _, a in ipairs(default) do
  log.raw("dev", "D %s %s\n", vim.inspect(a.key), a.action)
end

local active = api.config.mappings.active()
for _, a in ipairs(active) do
  log.raw("dev", "A %s %s\n", vim.inspect(a.key), a.action)
end
```

**Comment by alex-courtis** [comment]
Add `[NvimTree] "No view.mappings.list for on_attach generation."` message when using defaults / no legacy mappings.

**Comment by alex-courtis** [comment]
I think we are finally ready to release this:
* on_attach is generated from existing view.mappings - silent migration
* new HELP
* :help updated
* `:NvimTreeGenerateOnAttach` added
* wiki [Migrating To on_attach](https://github.com/nvim-tree/nvim-tree.lua/wiki/Migrating-To-on_attach)

@gegoune I would be most grateful if you could review the docs, help, wiki etc and sanity test the migration wiki steps with your configuration.

File: doc/nvim-tree-lua.txt [hunk 1]
```diff
@@ -1483,121 +1600,6 @@ Examples:
 
           ----
 <
-Mouse support defined in |KeyBindings|
-
-Single left mouse mappings can be achieved via `<LeftRelease>`.
-
-Single right / middle mouse mappings will requre changes to |mousemodel| or |mouse|.
-
-DEFAULT MAPPINGS                                     *nvim-tree-default-mappings*
```
Comment (alex-courtis) [inline doc/nvim-tree-lua.txt:1492]: Removed these as they are problematic to generate; DEFAULT_ON_ATTACH is quite readable.

**Comment by gegoune** [comment]
Noticed that

```lua
  vim.keymap.set('n', 'D', '', { buffer = bufnr })
  vim.keymap.del('n', 'D', { buffer = bufnr })
```

Is it easy enough to detect and remove that pair altogether? If not mentioning it on wiki might be enough.

**Comment by alex-courtis** [comment]
> Is it easy enough to detect and remove that pair altogether? If not mentioning it on wiki might be enough.

That one is annoying. I failed at reliably detecting existing mappings so just gave up and mapped/unmapped. I'll add a comment in the generated file.

**Comment by alex-courtis** [comment]
> We could turn last argument (`opts`) into some helper function like

Heh... I have a helper like that in my config.

Yes. That will make the default and generated much easier to read.

**Comment by alex-courtis** [comment]
> Also, mappings with `action_cb` did not get their code copied over. I can see that you added some skeleton there so it's good enough for me, probably quite hard to parse it out of configuration file.

Yeah... we tried a few tricks but gave up... even getting the function name reliably is an issue.

**Comment by alex-courtis** [comment]
Thank you so much for reviewing... this one is horrible.

I'll try and get the updates in tomorrow.

**Comment by gegoune** [comment]
There could be a comment on top of generated function saying something along those lines: "Attention: body of this function was automatically generated. Although care was taken to ensure (?)its correctness and completeness(/?) user's review is required. Please check for following issues in auto generated content: <list here>". 

I have been using that branch entire week. It just works.

**Comment by alex-courtis** [comment]
> I have been using that branch entire week. It just works.

This is great news.

I was tempted to try and revert some legacy bits to current, now that you've made on_attach much simpler, however working code should be used.

Finishing last little tasks and releasing...
